### ISSUE TITLE
Add support for strings in CGen backend

### ISSUE THREAD

**Issue by lkuper** [IntelLabs/ParallelAccelerator.jl#49]
Paul and I began making some headway on supporting the `ASCIIString` type with https://github.com/IntelLabs/ParallelAccelerator.jl/commit/14b57e7fd5b1f065a0ac4617c8a653565a47a725.  We still don't support passing strings to, or returning strings, from `@acc`-annotated functions.  See `test/strings.jl` for what's supported so far.

See also #48.

**Comment by ninegua** [comment]
Where do we stand on this? We have very basic support for string literals, as the following runs fine:
```
using ParallelAccelerator
@acc f()=println("Hello world!")
f()
@acc g(x)=println(x)
g("Hello world!")
```

How about we close this ticket until there is demand to go further than the above (e.g., string functions)?

**Comment by ninegua** [comment]
Ok, I worked on it a bit more, revived the `ASCIIString` class in C++ (so that unlike char*, it will automatically work with other Julia functions such as length since in Julia ASCIIString is just a normal data type with `data` field that is an UInt8 array), and implemented Julia/C conversions in commit 6189d00300d9b5233b551760a2382113de3cc6a5). The improved test case [`test/strings.jl`](https://github.com/IntelLabs/ParallelAccelerator.jl/blob/master/test/strings.jl) now covers accessing string element, passing string as parameter, return string, and string concatenation. Other string functions may also work but are not tested, so please create new tickets if you need something and found it not working.

**Comment by ehsantn** [comment]
This change breaks HPAT since some C++ functions expect `char *`.

**Comment by ninegua** [comment]
It may be related to ccall with string arguments. I think it should be easy to fix. Can you give a small test case? @ehsantn

**Comment by ehsantn** [comment]
I think I can fix the HPAT issue easily. However, I get this error even for ParallelAccelerator string tests:

```
RROR: LoadError: LoadError: test error in expression: StringTest.test3()
ccall: could not find function new_ascii_string in library /home/etotoni/.julia/v0.4/ParallelAccelerator/src/../deps/libj2carray.so.1.0
 in ##_ppf3p14988_j2c_proxy#15074 at /home/etotoni/.julia/v0.4/ParallelAccelerator/src/driver.jl:359
 in f3 at /home/etotoni/.julia/v0.4/CompilerTools/src/OptFramework.jl:436
 in test3 at /home/etotoni/.julia/v0.4/ParallelAccelerator/test/strings.jl:63
 in anonymous at test.jl:92
 in do_test at test.jl:50
 in include at ./boot.jl:261
 in include_from_node1 at ./loading.jl:304
 in include at ./boot.jl:261
 in include_from_node1 at ./loading.jl:304
 in process_options at ./client.jl:280
 in _start at ./client.jl:378
 in ##_ppf3p14988_j2c_proxy#15074 at /home/etotoni/.julia/v0.4/ParallelAccelerator/src/driver.jl:359
 in f3 at /home/etotoni/.julia/v0.4/CompilerTools/src/OptFramework.jl:436
 in test3 at /home/etotoni/.julia/v0.4/ParallelAccelerator/test/strings.jl:63
 in anonymous at test.jl:92
 in do_test at test.jl:50
 in include at ./boot.jl:261
 in include_from_node1 at ./loading.jl:304
 in include at ./boot.jl:261
 in include_from_node1 at ./loading.jl:304
 in process_options at ./client.jl:280
 in _start at ./client.jl:378
while loading /home/etotoni/.julia/v0.4/ParallelAccelerator/test/strings.jl, in expression starting on line 77
while loading /home/etotoni/.julia/v0.4/ParallelAccelerator/test/runtests.jl, in expression starting on line 44

```

**Comment by ninegua** [comment]
Likely you need to run `build.sh` again, or `julia -e 'Pkg.build("ParallelAccelerator")'

**Comment by ehsantn** [comment]
You are right. `Pkg.update()` doesn't do it for some reason.

**Comment by ehsantn** [comment]
Clang build has a problem:
```
/Users/travis/.julia/v0.4/ParallelAccelerator/src/../deps/generated/cgen_output49.cpp:30:11: error: character too large for enclosing character literal type
GenSym0 = '�';
```

**Comment by ninegua** [comment]
The clang issue has been fixed in a5a68b038e488a26e1de5fcfe64f41dc5e57f99e.

**Comment by ehsantn** [comment]
Thanks.
